packages/ast-spec/src/declaration/TSDeclareFunction/spec.ts

```diff
@@ -1,11 +1,35 @@
 import type { AST_NODE_TYPES } from '../../ast-node-types';
 import type { FunctionBase } from '../../base/FunctionBase';
-import type { BlockStatement } from '../../statement/BlockStatement/spec';
 
-// TODO(#1852) - async + declare are semantically invalid together
-export interface TSDeclareFunction extends FunctionBase {
+interface TSDeclareFunctionBase extends FunctionBase {
   type: AST_NODE_TYPES.TSDeclareFunction;
-  body: BlockStatement | undefined;
+  /**
+   * TS1183: An implementation cannot be declared in ambient contexts.
+   */
+  body: undefined;
+  /**
+   * Whether the declaration has `declare` modifier.
+   */
   declare: boolean;
   expression: false;
 }
+
+export interface TSDeclareFunctionWithDeclare extends TSDeclareFunctionBase {
+  /**
+   * TS1040: 'async' modifier cannot be used in an ambient context.
+   */
+  async: false;
+  declare: true;
+  /**
+   * TS1221: Generators are not allowed in an ambient context.
+   */
+  generator: false;
+}
+
+export interface TSDeclareFunctionNoDeclare extends TSDeclareFunctionBase {
+  declare: false;
+}
```

Comment:
```ts
function* foo(): Generator<never, void, unknown>;
//      ^ TS1222: An overload signature cannot be declared as a generator.
function* foo(): Generator<never, void, unknown> { }
```
[playground](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.ts&code=GYVwdgxgLglg9mAVAAmHOAKAlALmQcQFMxCAnAQyjlIB4SA3MgGmXrhgBMXwBrMOAO5gAfAG4AUKEiwEKNJlwFiZStTqFGpFm07cwfQSOQBvZAF8gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

Looks like we can exclude and crash on this case as well.

Reply:
Huh, TIL.